zellij-utils/src/input/mouse.rs

```diff
@@ -2,102 +2,135 @@ use serde::{Deserialize, Serialize};
 
 use crate::position::Position;
 
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, Deserialize, Serialize)]
+/// A mouse event can have any number of buttons (including no
+/// buttons) pressed or released.
+pub struct MouseButtons {
+    pub left: bool,
+    pub right: bool,
+    pub middle: bool,
+    pub wheel_up: bool,
+    pub wheel_down: bool,
+}
+
 /// A mouse related event
 #[derive(Debug, Copy, Clone, PartialEq, Eq, Deserialize, Serialize)]
 pub enum MouseEvent {
     /// A mouse button was pressed.
     ///
     /// The coordinates are zero-based.
-    Press(MouseButton, Position),
+    Press(MouseButtons, Position),
     /// A mouse button was released.
     ///
     /// The coordinates are zero-based.
-    Release(Position),
+    Release(MouseButtons, Position),
     /// A mouse button is held over the given coordinates.
     ///
     /// The coordinates are zero-based.
-    Hold(Position),
+    Motion(MouseButtons, Position),
 }
 
-impl From<termwiz::input::MouseEvent> for MouseEvent {
-    fn from(event: termwiz::input::MouseEvent) -> Self {
-        #[allow(clippy::if_same_then_else)]
-        if event
-            .mouse_buttons
-            .contains(termwiz::input::MouseButtons::LEFT)
+impl MouseButtons {
+    fn termwiz_mouse_buttons(mut self, button_bits: &termwiz::input::MouseButtons) -> MouseButtons {
+        self.left = button_bits.contains(termwiz::input::MouseButtons::LEFT);
+        self.right = button_bits.contains(termwiz::input::MouseButtons::RIGHT);
+        self.middle = button_bits.contains(termwiz::input::MouseButtons::MIDDLE);
+        self.wheel_up = button_bits.contains(termwiz::input::MouseButtons::VERT_WHEEL)
+            && button_bits.contains(termwiz::input::MouseButtons::WHEEL_POSITIVE);
+        self.wheel_down = button_bits.contains(termwiz::input::MouseButtons::VERT_WHEEL)
+            && !button_bits.contains(termwiz::input::MouseButtons::WHEEL_POSITIVE);
+
+        self
+    }
+}
+
+impl MouseEvent {
+    #[allow(unused)]
+    pub fn from_termwiz(mut old_buttons: MouseButtons, event: termwiz::input::MouseEvent) -> Self {
+        // We use the state of old_buttons vs new_buttons to determine
+        // if this event is a Press, Release, or Motion.  This is an
+        // unfortunate side effect of the pre-SGR-encoded X10 mouse
+        // protocol design in which release events don't carry
+        // information about WHICH button(s) were released, so we have
+        // to maintain a wee bit of state in between events.
```

Comment:
Honestly, even this is much better than what we've been doing up until now with the HeldMouse stuff.